src/common/checksum.c

```diff
@@ -652,37 +653,47 @@ calc_csum_sv(struct daos_csummer *obj, d_sg_list_t *sgl, size_t rec_len,
 		D_ASSERT(singv_lo->cs_bytes > 0 &&
 			 singv_lo->cs_bytes < rec_len);
 		D_ASSERT(singv_lo->cs_nr > 1);
-		last_idx = rec_len / singv_lo->cs_bytes;
-		D_ASSERT(last_idx >= 1);
+		data_tgt_nr = roundup(rec_len, singv_lo->cs_bytes) /
+			      singv_lo->cs_bytes;
+		D_ASSERT(data_tgt_nr >= 2);
+		last_size = rec_len - (data_tgt_nr - 1) * singv_lo->cs_bytes;
+		bytes_for_csum = singv_lo->cs_bytes;
 		if (singv_idx == -1) {
 			D_ASSERT(csums->cs_nr == singv_lo->cs_nr);
-			last_size = rec_len - last_idx * singv_lo->cs_bytes;
 		} else {
 			D_ASSERT(csums->cs_nr == 1);
 			/* skip to the sgl location of singv_idx, the last
 			 * data cell possibly with less valid bytes, and
 			 * followed by parity cells.
 			 */
-			if (singv_idx <= last_idx)
-				skip_size = singv_idx * singv_lo->cs_bytes;
+			if (singv_idx < data_tgt_nr)
+				singv_off = singv_idx * singv_lo->cs_bytes;
 			else
-				skip_size = rec_len + (singv_idx - last_idx) *
-						      singv_lo->cs_bytes;
-			rc = daos_sgl_processor(sgl, false, &sgl_idx, skip_size,
+				singv_off = rec_len +
+					    (singv_idx - data_tgt_nr) *
+					    singv_lo->cs_bytes;
+			rc = daos_sgl_processor(sgl, false, &sgl_idx, singv_off,
 						NULL, NULL);
 			if (rc)
 				return rc;
+			if (singv_idx == data_tgt_nr - 1)
+				bytes_for_csum = last_size;
 		}
-		bytes_for_csum = singv_lo->cs_bytes;
 	} else {
 		D_ASSERT(csums->cs_nr == 1);
 		bytes_for_csum = rec_len;
 	}
 
 	csum_buf_len = bytes_for_csum;
 	for (idx = 0; idx < csums->cs_nr; idx++) {
-		if (idx == last_idx)
-			csum_buf_len = last_size;
+		if (singv_lo != NULL && singv_lo->cs_even_dist == 1 &&
+		    singv_idx == -1) {
+			if (idx == data_tgt_nr - 1)
+				csum_buf_len = last_size;
```

Comment:
csum_buf_len is confusing to me. Sounds like it's the buffer length of the checksums, but looks like it is the number of bytes of the data that will be used to calculate the checksum? Is that right?

Reply:
correct. hmm, agreed the name looks confuse, let me rename it...